### DISCUSSION TITLE
can we make module "cryptography" optional?

### DISCUSSION THREAD

**Discussion by sanderjo** [sabnzbd/sabnzbd#2998] [Ideas]
Can we make module "cryptography" optional?

Reasons:
* on non-mainstream platforms (OS, CPU) it needs compiling, which needs cargo / rust, which is big, and which is hard on non-mainstream platforms
* we use it to generate server.cert and server.key ... needed if you want SABnzbd to listen on HTTPS with self-signed. I tried: with copying in a server.cert and server.key generated by another SABnzbd: works too.
* optional use in rarfile. 

So: make the module optional. And if module not there: 
* in sabnzbd: check if there are server.cert and server.key. Only then provide HTTPS
* or use a more lightweight python module to generate self-signed certificate
* or provide a default server.cert and server.key with the installation
* (or generate a server.cert and server.key with openssl oneliner, on those non-mainstream platforms)

FWIW: the pre-built version of cryptography takes 12MB. Quite a lot for generating server.cert and server.key ... if used at all.

```
sander@zwarte:~/weg-venv$ du -h ./lib/python3.12/site-packages/cryptography/ | tail -1
12M	./lib/python3.12/site-packages/cryptography/
```

**Comment by sanderjo** [comment]
About "generate a server.cert and server.key with openssl oneliner"

In .sabnzbd/admin:

`openssl req -x509 -newkey rsa:4096 -keyout server.key -out server.cert -sha256 -days 3650 -nodes -subj "/C=XX/ST=StateName/L=CityName/O=SABnzbd/OU=SABnzbd/CN=SABnzbd-selfsigned"`

and it works: SABnzbd can serve via HTTPS.

![image](https://github.com/user-attachments/assets/2d1c8fa8-b624-4163-ba77-b455d1ceba4d)

Of course as bad as all self-signed. But: a one-liner with openssl. No 12MB module, no compiling with Rust

**Comment by sanderjo** [comment]
I don't know if this is pythonic, but I saw this:

`pip install -r requirements.txt -r optional-requirements.txt
`
... also put in there the json modules.

**Comment by Safihre** [comment]
The main problem is that it's required to read encrypted Rar files. So it's essential for our Encrypted Rar detection and reliability of Direct Unpack and regular Unpack.

We can't get rid of it...

**Reply by sanderjo** [reply]
the rarfile part already seems optional? Or am I reading this wrong?

https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/utils/rarfile.py#L108-L130

**Reply by Safihre** [reply]
Yes, it won't crash but just disable the reading of encrypted rars.
The fallback for cryptography is another binary package that needs compiling.
